sbioapputils/app_runner/workflow_utils.py

```diff
@@ -0,0 +1,136 @@
+    name, stages, parameters = _parse_workflow()
+    for key, value in request['input_files'].items():
+        request[key] = value
+    request['job_id'] = job_id
+
+    # Check all required parameters are provided
+    no_default = []
+    default = {}
+    no_type = []
+    wrong_data_types = []
+    invalid_value = []
+
+    for key in parameters.keys():
+
+        # Check if type is present
+        if not parameters[key].get('type'):
+            no_type.append(key)
+        else:
+            # Check if type is correct
+            try:
+                if not isinstance(request[key], parameters[key]['type']):
+                    wrong_data_types.append(key)
+            # On error if config[key] not present
+            except KeyError:
+                pass
+
+        # Check if default is present
+        if not parameters[key].get('default'):
+            no_default.append(key)
+        else:
+            default[key] = parameters[key]['default']
+            # Set defaults if not present
+            if not request.get(key):
+                request[key] = parameters[key]['default']
+
+        if parameters[key].get('user_defined'):
+            if parameters[key]['user_defined'] == 'True':
+                if parameters[key]['type'] in ['int', 'float']:
+                    # Check between min and max
+                    if parameters[key].get('max_value'):
+                        if (request[key] > parameters[key]['max_value']):
+                            invalid_value.append(key)
+                    if parameters[key].get('min_value'):
+                        if (request[key] < parameters[key]['min_value']):
+                            invalid_value.append(key)
+
+                elif parameters[key]['type'] == 'str':
+                    if parameters[key].get('from_data'):
+                        if parameters[key]['from_data'] == 'True':
+                            dropdown = False
+                        else:
+                            dropdown = True
+                    else:
+                        dropdown = True
+
+                    # Category settings
+                    if dropdown:
+                        if request[key] not in parameters[key]['options']:
+                            invalid_value.append(key)
+
+        # Create directory if needed
+        try:
+            if str(request[key])[-1] == '/':
+                if not os.path.exists(request[key]):
+                    os.mkdir(request[key])
+                    #print("Directory '% s' created" % request[key])
+        except KeyError:
+            pass
+
+    if not all(param in request.keys() for param in no_default):
+        raise Exception(f"These required parameters are not specified: {list(set(no_default) - set(request.keys()))}")
+
+    if not all(param in request.keys() for param in no_type):
+        raise Exception(f"These parameters have an incorrect data type: {list(set(wrong_data_types) - set(request.keys()))}")
```

Comment:
why isn't it just the wrong_data_types but the difference between wrong_data_types and all keys?

Reply:
I'm not sure why I did it that way. It does not seem to be correct, so I've changed it. Same for line 100.